src/charge/modifiers/Once.php

```diff
@@ -0,0 +1,141 @@
+<?php declare(strict_types=1);
+
+namespace hiqdev\php\billing\charge\modifiers;
+
+use DateTimeImmutable;
+use hiqdev\php\billing\action\ActionInterface;
+use hiqdev\php\billing\charge\ChargeInterface;
+use hiqdev\php\billing\charge\modifiers\addons\MonthPeriod;
+use hiqdev\php\billing\charge\modifiers\addons\Period;
+use hiqdev\php\billing\charge\modifiers\addons\YearPeriod;
+use hiqdev\php\billing\charge\modifiers\exception\OnceException;
+
+/**
+ * 1. API:
+ * - once.per('1 year') – bill every month that matches the month of sale of the object
+ * - once.per('3 months') – bill every third month, starting from the month of sale of the object
+ * - once.per('day'), once.per('1.5 months') – throws an interpret-time exception, a value must NOT be a fraction of the month
+ * 2. In months where the formula should NOT bill, it should produce a ZERO charge.
+ * 3. If the sale is re-opened, the formula starts over.
+ * 4. It should throw exception for overuse prices.
+ */
+class Once extends Modifier
+{
+    private const MONTHS_IN_YEAR = 12;
+
+    public function per(string $interval): self
+    {
+        return $this->addAddon('per', $this->createInterval($interval));
+    }
+
+    private function createInterval(string $interval): Period
+    {
+        if ($this->isInvalidInterval($interval)) {
+            throw new OnceException("The interval cannot be a fraction.");
+        }
+
+        if ($this->isSupportedInterval($interval)) {
+            return Period::fromString($interval);
+        }
+
+        throw new OnceException("Invalid interval. Supported: whole months or years.");
+    }
+
+    private function isInvalidInterval(string $interval): bool
+    {
+        return preg_match('/(\d+(\.\d+)?)\s*([a-z]+)/', $interval, $matches)
+            && (
+                floatval($matches[1]) != intval($matches[1]) ||   // Check for fractional values
+                !preg_match('/months?|years?/', $matches[3])      // Ensure only 'months' or 'years' are allowed
+            );
+    }
+
+    private function isSupportedInterval(string $interval): bool
+    {
+        return str_contains($interval, 'month') || str_contains($interval, 'year');
+    }
+
+    public function getPer(): ?Period
+    {
+        return $this->getAddon('per');
+    }
+
+    public function modifyCharge(?ChargeInterface $charge, ActionInterface $action): array
+    {
+        $period = $this->getPer();
+
+        $this->assertPeriod($period);
+        $this->assertCharge($charge);
+        $this->assertNoOverusePricing($charge);
+
+        // Apply the charge if applicable based on the action and interval period
+        if ($this->isApplicable($action, $period)) {
+            return [$charge];
+        }
+
+        // Return zero charge if the period is not applicable
+        return [];
```

Comment:
It's a NULL charge ;)

Please, use ChargeDerivative to create a charge with 0 sum. This way we create an uninterrupted chain of charges in a timeline and know that a particular action was processed each month.